File: Raven.Client.Lightweight/Connection/HttpJsonRequestFactory.cs
using System;
using System.Collections.Specialized;
using System.Net;
using System.Threading;
using Raven.Abstractions.Extensions;
using Raven.Client.Connection.Profiling;
using Raven.Client.Util;
using Raven.Json.Linq;

namespace Raven.Client.Connection
{
	///<summary>
	/// Create the HTTP Json Requests to the RavenDB Server
	/// and manages the http cache
	///</summary>
	public class HttpJsonRequestFactory : IDisposable
	{
		/// <summary>
		/// Advanced: Don't set this unless you know what you are doing!
		/// 
		/// Enable using basic authentication using http
		/// By default, RavenDB only allows basic authentication over HTTPS, setting this property to true
		/// will instruct RavenDB to make unsecure calls (usually only good for testing / internal networks).
		/// </summary>
		public bool EnableBasicAuthenticationOverUnsecureHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers { get; set; }

		/// <summary>
		/// Occurs when a json request is created
		/// </summary>
		public event EventHandler<WebRequestEventArgs> ConfigureRequest = delegate { };

		/// <summary>
		/// Occurs when a json request is completed
		/// </summary>
		public event EventHandler<RequestResultArgs> LogRequest = delegate { };

		/// <summary>
		/// Invoke the LogRequest event
		/// </summary>
		internal void InvokeLogRequest(IHoldProfilingInformation sender, Func<RequestResultArgs> generateRequentResult)
		{
			var handler = LogRequest;
			if (handler != null) 
				handler(sender, generateRequentResult());
		}

		private readonly int maxNumberOfCachedRequests;
		private SimpleCache<CachedRequest> cache;

		internal int NumOfCachedRequests;

		/// <summary>
		/// Creates the HTTP json request.
		/// </summary>
		public HttpJsonRequest CreateHttpJsonRequest(CreateHttpJsonRequestParams createHttpJsonRequestParams)
		{
			if (disposed)
				throw new ObjectDisposedException(typeof(HttpJsonRequestFactory).FullName);
			var request = new HttpJsonRequest(createHttpJsonRequestParams, this)
			{
				ShouldCacheRequest = createHttpJsonRequestParams.Convention.ShouldCacheRequest(createHttpJsonRequestParams.Url)
			};

			if (request.ShouldCacheRequest && createHttpJsonRequestParams.Method == "GET" && !DisableHttpCaching)
			{
				var cachedRequestDetails = ConfigureCaching(createHttpJsonRequestParams.Url, request.webRequest.Headers.Set);
				request.CachedRequestDetails = cachedRequestDetails.CachedRequest;
				request.SkipServerCheck = cachedRequestDetails.SkipServerCheck;
			}
			ConfigureRequest(createHttpJsonRequestParams.Owner, new WebRequestEventArgs {Request = request.webRequest});
			return request;
		}

		internal CachedRequestOp ConfigureCaching(string url, Action<string, string> setHeader)
		{
			var cachedRequest = cache.Get(url);
			if (cachedRequest == null)
				return new CachedRequestOp { SkipServerCheck = false };
			bool skipServerCheck = false;
			if (AggressiveCacheDuration != null)
			{
				var duration = AggressiveCacheDuration.Value;
				if(duration.TotalSeconds > 0)
					setHeader("Cache-Control", "max-age=" + duration.TotalSeconds);

				if ((DateTimeOffset.Now - cachedRequest.Time) < duration) // can serve directly from local cache
					skipServerCheck = true;
			}

			setHeader("If-None-Match", cachedRequest.Headers["ETag"]);
			return new CachedRequestOp { SkipServerCheck = skipServerCheck, CachedRequest = cachedRequest };
		}


		/// <summary>
		/// Reset the number of cached requests and clear the entire cache
		/// Mostly used for testing
		/// </summary>
		public void ResetCache()
		{
			if (cache != null)
				cache.Dispose();

			cache = new SimpleCache<CachedRequest>(maxNumberOfCachedRequests);
			NumOfCachedRequests = 0;
		}



		/// <summary>
		/// The number of requests that we got 304 for 
		/// and were able to handle purely from the cache
		/// </summary>
		public int NumberOfCachedRequests
		{
			get { return NumOfCachedRequests; }
		}

		/// <summary>
		/// Determine whether to use compression or not 
		/// </summary>
		public bool DisableRequestCompression { get; set; }

		/// <summary>
		/// default ctor
		/// </summary>
		/// <param name="maxNumberOfCachedRequests"></param>
		public HttpJsonRequestFactory(int maxNumberOfCachedRequests)
		{
			this.maxNumberOfCachedRequests = maxNumberOfCachedRequests;
			ResetCache();
		}

#if !NET35
		///<summary>
		/// The aggressive cache duration
		///</summary>
		public TimeSpan? AggressiveCacheDuration
		{
			get { return aggressiveCacheDuration.Value; }
			set { aggressiveCacheDuration.Value = value; }
		}

		/// <summary>
		/// Disable the HTTP caching
		/// </summary>
		public bool DisableHttpCaching
		{
			get { return disableHttpCaching.Value; }
			set { disableHttpCaching.Value = value; }
		}

		private readonly ThreadLocal<TimeSpan?> aggressiveCacheDuration = new ThreadLocal<TimeSpan?>(() => null);

		private readonly ThreadLocal<bool> disableHttpCaching = new ThreadLocal<bool>(() => false);
#else
		[ThreadStatic] private static TimeSpan? aggressiveCacheDuration;
		[ThreadStatic] private static bool disableHttpCaching;


		
		/// <summary>
		/// Disable the HTTP caching
		/// </summary>
		public bool DisableHttpCaching
		{
			get { return disableHttpCaching; }
			set { disableHttpCaching = value; }
		}

		///<summary>
		/// The aggressive cache duration
		///</summary>
		public TimeSpan? AggressiveCacheDuration
		{
			get { return aggressiveCacheDuration; }
			set { aggressiveCacheDuration = value; }
		}
#endif
		private volatile bool disposed;

		internal RavenJToken GetCachedResponse(HttpJsonRequest httpJsonRequest)
		{
			if (httpJsonRequest.CachedRequestDetails == null)
				throw new InvalidOperationException("Cannot get cached response from a request that has no cached information");
			httpJsonRequest.ResponseStatusCode = HttpStatusCode.NotModified;
			httpJsonRequest.ResponseHeaders = new NameValueCollection(httpJsonRequest.CachedRequestDetails.Headers);
			IncrementCachedRequests();
			return httpJsonRequest.CachedRequestDetails.Data.CloneToken();
		}

		internal void IncrementCachedRequests()
		{
			Interlocked.Increment(ref NumOfCachedRequests);
		}

		internal void CacheResponse(string url, RavenJToken data, NameValueCollection headers)
		{
			if (string.IsNullOrEmpty(headers["ETag"])) 
				return;

			var clone = data.CloneToken();
			clone.EnsureSnapshot();
			cache.Set(url, new CachedRequest
			{
				Data = clone,
				Time = DateTimeOffset.Now,
				Headers = new NameValueCollection(headers)
			});
		}

		/// <summary>
		/// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
		/// </summary>
		/// <filterpriority>2</filterpriority>
		public void Dispose()
		{
			if (disposed)
				return;
		    disposed = true;
			cache.Dispose();
#if !NET35
			aggressiveCacheDuration.Dispose();
			disableHttpCaching.Dispose();
#endif
		}

		internal void UpdateCacheTime(HttpJsonRequest httpJsonRequest)
		{
			if (httpJsonRequest.CachedRequestDetails == null)
				throw new InvalidOperationException("Cannot update cached response from a request that has no cached information");
			httpJsonRequest.CachedRequestDetails.Time = DateTimeOffset.Now;
		}

		/// <summary>
		/// Disable all caching within the given scope
		/// </summary>
		public IDisposable DisableAllCaching()
		{
			var oldAgressiveCaching = AggressiveCacheDuration;
			var oldHttpCaching = DisableHttpCaching;

			AggressiveCacheDuration = null;
			DisableHttpCaching = true;

			return new DisposableAction(() =>
			{
				AggressiveCacheDuration = oldAgressiveCaching;
				DisableHttpCaching = oldHttpCaching;
			});
		}
	}
}

File: Raven.Client.Java/src/main/java/net/ravendb/client/connection/implementation/HttpJsonRequestFactory.java
package net.ravendb.client.connection.implementation;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

import net.ravendb.abstractions.basic.EventHandler;
import net.ravendb.abstractions.basic.EventHelper;
import net.ravendb.abstractions.closure.Action2;
import net.ravendb.abstractions.connection.WebRequestEventArgs;
import net.ravendb.abstractions.data.Constants;
import net.ravendb.abstractions.data.HttpMethods;
import net.ravendb.abstractions.json.linq.RavenJToken;
import net.ravendb.client.connection.CachedRequest;
import net.ravendb.client.connection.CachedRequestOp;
import net.ravendb.client.connection.CreateHttpJsonRequestParams;
import net.ravendb.client.connection.profiling.IHoldProfilingInformation;
import net.ravendb.client.connection.profiling.RequestResultArgs;
import net.ravendb.client.extensions.MultiDatabase;
import net.ravendb.client.util.SimpleCache;

import org.apache.commons.lang.StringUtils;
import org.apache.http.HttpRequest;
import org.apache.http.HttpStatus;
import org.apache.http.config.SocketConfig;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.impl.client.StandardHttpRequestRetryHandler;
import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;


/**
 * Create the HTTP Json Requests to the RavenDB Server
 * and manages the http cache
 */
public class HttpJsonRequestFactory implements AutoCloseable {

  private CloseableHttpClient httpClient;

  private List<EventHandler<WebRequestEventArgs>> configureRequest = new ArrayList<>();

  private List<EventHandler<RequestResultArgs>> logRequest = new ArrayList<>();

  private final int maxNumberOfCachedRequests;
  private SimpleCache cache;
  protected AtomicInteger numOfCachedRequests = new AtomicInteger();
  protected int numOfCacheResets;
  private boolean disableRequestCompression;
  private boolean enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers;
  private ThreadLocal<Long> aggressiveCacheDuration = new ThreadLocal<>(); // in milis
  private ThreadLocal<Boolean> disableHttpCaching = new ThreadLocal<>();
  private volatile boolean disposed;
  private ThreadLocal<Long> requestTimeout=  new ThreadLocal<>();// in milis

  public HttpJsonRequestFactory(int maxNumberOfCachedRequests) {
    super();

    PoolingHttpClientConnectionManager cm = new PoolingHttpClientConnectionManager();
    cm.setDefaultMaxPerRoute(10);
    this.httpClient = HttpClients.custom().setConnectionManager(cm).setRetryHandler(new StandardHttpRequestRetryHandler(0, false))
      .setDefaultSocketConfig(SocketConfig.custom().setTcpNoDelay(true).build()).
      build();
    this.maxNumberOfCachedRequests = maxNumberOfCachedRequests;
    resetCache();
  }

  public void addConfigureRequestEventHandler(EventHandler<WebRequestEventArgs> event) {
    configureRequest.add(event);
  }


  public void addLogRequestEventHandler(EventHandler<RequestResultArgs> event) {
    logRequest.add(event);
  }

  /**
   * @return the httpClient
   */
  public CloseableHttpClient getHttpClient() {
    return httpClient;
  }

  public void cacheResponse(String url, RavenJToken data, Map<String, String> headers) {
    if (StringUtils.isEmpty(headers.get("ETag"))) {
      return;
    }

    RavenJToken clone = data.cloneToken();
    clone.ensureCannotBeChangeAndEnableShapshotting();

    cache.set(url, new CachedRequest(clone, new Date(), new HashMap<>(headers), MultiDatabase.getDatabaseName(url), false));

  }

  @Override
  public void close() throws Exception {
    if (disposed) {
      return ;
    }
    disposed = true;
    cache.close();
    httpClient.close();

  }

  public CachedRequestOp configureCaching(String url, Action2<String, String> setHeader) {
    CachedRequest cachedRequest = cache.get(url);
    if (cachedRequest == null) {
      return new CachedRequestOp(null, false);
    }
    boolean skipServerCheck = false;
    if (getAggressiveCacheDuration() != null) {
      long totalSeconds = getAggressiveCacheDuration() / 1000;
      if (totalSeconds > 0) {
        setHeader.apply("Cache-Control", "max-age=" + totalSeconds);
      }

      if (cachedRequest.isForceServerCheck() == false && (new Date().getTime() - cachedRequest.getTime().getTime() < getAggressiveCacheDuration())) { //can serve directly from local cache
        skipServerCheck = true;
      }
      cachedRequest.setForceServerCheck(false);
    }
    setHeader.apply("If-None-Match", cachedRequest.getHeaders().get("ETag"));
    return new CachedRequestOp(cachedRequest, skipServerCheck);
  }

  private static class SetHeader implements Action2<String, String> {

    private HttpRequest request;

    public SetHeader(HttpRequest request) {
      this.request = request;
    }

    @Override
    public void apply(String headerName, String value) {
      request.addHeader(headerName, value);
    }

  }

  public HttpJsonRequest createHttpJsonRequest(CreateHttpJsonRequestParams createHttpJsonRequestParams) {
    if (disposed) {
      throw new IllegalStateException("Object was disposed!");
    }

    HttpJsonRequest request = new HttpJsonRequest(createHttpJsonRequestParams, this);
    request.setShouldCacheRequest(createHttpJsonRequestParams.isAvoidCachingRequest() == false
        && createHttpJsonRequestParams.getConvention().shouldCacheRequest(createHttpJsonRequestParams.getUrl()));

    if (request.getShouldCacheRequest() && createHttpJsonRequestParams.getMethod() == HttpMethods.GET && !getDisableHttpCaching()) {
      CachedRequestOp cachedRequestDetails = configureCaching(createHttpJsonRequestParams.getUrl(), new SetHeader(request.getWebRequest()));
      request.setCachedRequestDetails(cachedRequestDetails.getCachedRequest());
      request.setSkipServerCheck(cachedRequestDetails.isSkipServerCheck());
    }

    if (getRequestTimeout() != null) {
      request.setTimeout(getRequestTimeout());
    }

    EventHelper.invoke(configureRequest, createHttpJsonRequestParams.getOwner(), new WebRequestEventArgs(request.getWebRequest(), createHttpJsonRequestParams.getCredentials()));
    return request;
  }

  public AutoCloseable disableAllCaching() {
    final Long oldAggressiveCaching = getAggressiveCacheDuration();
    final Boolean oldHttpCaching = getDisableHttpCaching();

    setAggressiveCacheDuration(null);
    setDisableHttpCaching(true);

    return new AutoCloseable() {
      @Override
      public void close() throws Exception {
        setAggressiveCacheDuration(oldAggressiveCaching);
        setDisableHttpCaching(oldHttpCaching);
      }
    };
  }

  public void expireItemsFromCache(String db)
  {
    cache.forceServerCheckOfCachedItemsForDatabase(db);
    numOfCacheResets++;
  }

  public Long getAggressiveCacheDuration() {
    return aggressiveCacheDuration.get();
  }

  RavenJToken getCachedResponse(HttpJsonRequest httpJsonRequest, Map<String, String> additionalHeaders) {
    if (httpJsonRequest.getCachedRequestDetails() == null) {
      throw new IllegalStateException("Cannot get cached response from a request that has no cached information");
    }
    httpJsonRequest.setResponseStatusCode(HttpStatus.SC_NOT_MODIFIED);
    httpJsonRequest.setResponseHeaders(new HashMap<>(httpJsonRequest.getCachedRequestDetails().getHeaders()));

    if (additionalHeaders != null && additionalHeaders.containsKey(Constants.RAVEN_FORCE_PRIMARY_SERVER_CHECK)) {
      httpJsonRequest.getResponseHeaders().put(Constants.RAVEN_FORCE_PRIMARY_SERVER_CHECK, additionalHeaders.get(Constants.RAVEN_FORCE_PRIMARY_SERVER_CHECK));
    }

    incrementCachedRequests();
    return httpJsonRequest.getCachedRequestDetails().getData().cloneToken();
  }

  /**
   * The number of currently held requests in the cache
   * @return
   */
  public int getCurrentCacheSize() {
    return cache.getCurrentSize();
  }

  public boolean getDisableHttpCaching() {
    Boolean value = disableHttpCaching.get();
    if (value == null) {
      return false;
    }
    return disableHttpCaching.get();
  }


  /**
   * @return the numOfCachedRequests
   */
  public int getNumOfCachedRequests() {
    return numOfCachedRequests.get();
  }

  /**
   * @return the numOfCacheResets
   */
  public int getNumOfCacheResets() {
    return numOfCacheResets;
  }



  public void incrementCachedRequests() {
    numOfCachedRequests.incrementAndGet();
  }

  public void invokeLogRequest(IHoldProfilingInformation sender, RequestResultArgs requestResult) {
    EventHelper.invoke(logRequest, sender, requestResult);
  }

  /**
   * @return the disableRequestCompression
   */
  public boolean isDisableRequestCompression() {
    return disableRequestCompression;
  }

  /**
   * @return the enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers
   */
  public boolean isEnableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers() {
    return enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers;
  }

  public void removeConfigureRequestEventHandler(EventHandler<WebRequestEventArgs> event) {
    configureRequest.remove(event);
  }

  public void removeLogRequestEventHandler(EventHandler<RequestResultArgs> event) {
    logRequest.remove(event);
  }



  public void resetCache() {
    if (cache != null) {
      try {
        cache.close();
      } catch (Exception e) { /*ignore */ }
    }
    cache = new SimpleCache(maxNumberOfCachedRequests);
    numOfCachedRequests = new AtomicInteger();
  }

  public void setAggressiveCacheDuration(Long value) {
    aggressiveCacheDuration.set(value);
  }

  public void setDisableHttpCaching(Boolean value) {
    disableHttpCaching.set(value);
  }

  /**
   * @param disableRequestCompression the disableRequestCompression to set
   */
  public void setDisableRequestCompression(boolean disableRequestCompression) {
    this.disableRequestCompression = disableRequestCompression;
  }

  /**
   *  Advanced: Don't set this unless you know what you are doing!
   *  Enable using basic authentication using http
   *  By default, RavenDB only allows basic authentication over HTTPS, setting this property to true
   *  will instruct RavenDB to make unsecured calls (usually only good for testing / internal networks).
   * @param enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers the enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers to set
   */
  public void setEnableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers(
      boolean enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers) {
    this.enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers = enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers;
  }

  /**
   * @param numOfCacheResets the numOfCacheResets to set
   */
  public void setNumOfCacheResets(int numOfCacheResets) {
    this.numOfCacheResets = numOfCacheResets;
  }

  protected void updateCacheTime(HttpJsonRequest httpJsonRequest) {
    if (httpJsonRequest.getCachedRequestDetails() == null) {
      throw new IllegalStateException("Cannot update cached response from a request that has no cached information");
    }
    httpJsonRequest.getCachedRequestDetails().setTime(new Date());
  }

  public Long getRequestTimeout() {
    return requestTimeout.get();
  }


  public void setRequestTimeout(Long requestTimeout) {
    this.requestTimeout.set(requestTimeout);
  }

}
